File: 174__Dungeon_Game.py
class Solution(object):
    def calculateMinimumHP(self, dungeon):
        """
        :type dungeon: List[List[int]]
        :rtype: int
        """
        if not dungeon or not dungeon[0]:
            return 0

        m, n = len(dungeon), len(dungeon[0])

        for i in range(m - 1, -1, -1):
            for j in range(n - 1, -1, -1):
                if i == m - 1 and j == n - 1:
                    if dungeon[i][j] < 0:
                        dungeon[i][j] = 1 - dungeon[i][j]
                    else:
                        dungeon[i][j] = 1
                else:
                    if i == m - 1:
                        tmp = dungeon[i][j + 1] - dungeon[i][j]
                    elif j == n - 1:
                        tmp = dungeon[i + 1][j] - dungeon[i][j]
                    else:
                        tmp = min(dungeon[i][j + 1], dungeon[i + 1][j]) - dungeon[i][j]
                    if tmp <= 0:
                        dungeon[i][j] = 1
                    else:
                        dungeon[i][j] = tmp
        return dungeon[0][0]

File: Java/174__Dungeon_Game.java
class Solution {
    public int calculateMinimumHP(int[][] dungeon) {
        if (dungeon == null || dungeon.length == 0 || dungeon[0].length == 0) return 0;
        int m = dungeon.length, n = dungeon[0].length;
        int[][] cache = new int[m + 1][n + 1];
        return helper(0, 0, dungeon, cache);
    }

    private int helper(int m, int n, int[][] dungeon, int[][] cache) {
        if (m == dungeon.length - 1 && n > dungeon[0].length - 1) return 1;
        if (m > dungeon.length - 1 && n == dungeon[0].length - 1) return 1;
        if (m > dungeon.length - 1 || n > dungeon[0].length - 1) return Integer.MAX_VALUE;
        if (cache[m][n] == 0) {
            int minHP = Math.min(helper(m + 1, n, dungeon, cache), helper(m, n + 1, dungeon, cache)) - dungeon[m][n];
            cache[m][n] = (minHP <= 0) ? 1 : minHP;
        }
        return cache[m][n];
    }
}
